drivers/crypto/crypto_it8xxx2_sha.c

```diff
@@ -0,0 +1,213 @@
+	0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1,
+	0x923f82a4, 0xab1c5ed5, 0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3,
+	0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174, 0xe49b69c1, 0xefbe4786,
+	0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
+	0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147,
+	0x06ca6351, 0x14292967, 0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13,
+	0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85, 0xa2bfe8a1, 0xa81a664b,
+	0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
+	0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a,
+	0x5b9cca4f, 0x682e6ff3, 0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208,
+	0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2
+};
+
+static void it8xxx2_sha256_init(bool init_k)
+{
+	int i;
+
+	chip_ctx.total_len = 0;
+	chip_ctx.w_input_index = 0;
+
+	/* Initialize hash values */
+	for (i = 0; i < ARRAY_SIZE(sha256_h0); i++)
+		chip_ctx.h[i] = sha256_h0[i];
+	/* Initialize array of round constants */
+	if (init_k) {
+		for (int i = 0; i < ARRAY_SIZE(sha256_k); i++)
+			chip_ctx.k[i] = sha256_k[i];
+	}
+}
+
+static void it8xxx2_sha256_module_calculation(void)
+{
+	struct it8xxx2_sha_regs *const sha_regs = IT8XXX2_SHA_REGS_BASE;
+	uint32_t key;
+
+	__unused uint8_t hash_ctrl;
+
+	/*
+	 * Since W field on it8xxx2 requires big-endian format, change byte
+	 * order before computing hash.
+	 */
+	for (int i = 0; i < SHA_SHA256_BLOCK_LEN_WORDS; i++) {
+		chip_ctx.w_sha[i] = sys_cpu_to_be32(chip_ctx.w_sha[i]);
+	}
+	/*
+	 * Global interrupt is disabled because the CPU cannot access memory
+	 * via the DLM (Data Local Memory) bus while HW module is computing
+	 * hash.
+	 */
+	key = irq_lock();
+	sha_regs->SHA_HASHCTRLR |= IT8XXX2_SHA_START_SHA256;
+	hash_ctrl = sha_regs->SHA_HASHCTRLR;
+	irq_unlock(key);
+
+	chip_ctx.w_input_index = 0;
+}
+
+static int it8xxx2_hash_handler(struct hash_ctx *ctx, struct hash_pkt *pkt,
+				bool finish)
+{
+	uint32_t rem_len = pkt->in_len, in_buf_idx = 0;
+
+	while (rem_len--) {
+		chip_ctx.w_input[chip_ctx.w_input_index++] =
+						pkt->in_buf[in_buf_idx++];
+		if (chip_ctx.w_input_index >= SHA_SHA256_BLOCK_LEN) {
+			it8xxx2_sha256_module_calculation();
+		}
+	}
+	chip_ctx.total_len += pkt->in_len;
+
+	if (finish) {
+		uint32_t *ob_ptr = (uint32_t *)pkt->out_buf;
+
+		/* Pre-processing (Padding) */
+		memset(&chip_ctx.w_input[chip_ctx.w_input_index],
+			0, SHA_SHA256_BLOCK_LEN - chip_ctx.w_input_index);
+		chip_ctx.w_input[chip_ctx.w_input_index] = 0x80;
+
+		if (chip_ctx.w_input_index >= 56) {
+			it8xxx2_sha256_module_calculation();
+			memset(&chip_ctx.w_input[chip_ctx.w_input_index],
+			0, SHA_SHA256_BLOCK_LEN - chip_ctx.w_input_index);
+		}
+		chip_ctx.w_sha[15] = sys_cpu_to_be32(chip_ctx.total_len * 8);
+		it8xxx2_sha256_module_calculation();
+
+		for (int i = 0; i < SHA_SHA256_HASH_LEN_WORDS; i++) {
+			ob_ptr[i] = sys_cpu_to_be32(chip_ctx.h[i]);
```

Comment:
This converts from the big-endian on the hardware to the little-endian of the CPU, so this should be:
```suggestion
			ob_ptr[i] = sys_be32_to_cpu(chip_ctx.h[i]);
```
`sys_cpu_to_be32` and `sys_be32_to_cpu()` both call `__bswap_32()`. but you should use the correct wrapper because it helps document the code.

Reply:
Done.